test/recipes/15-test_out_option.t

```diff
@@ -41,6 +41,11 @@ $rand_path .= "/test.pem";
 
 test_illegal_path($rand_path);
 test_legal_path('test.pem');
+SKIP: {
+    skip "'".File::Spec->devnull()."' may not be a workable null device for this cross compiled build", 1
+        unless (config('CROSS_COMPILE') // '') eq '';
```

Comment:
Except that when you build mingw on MSYS2, you have no CROSS_COMPILE prefix. It's just that [1st] gcc on your $PATH is [effectively] the cross-compiler. Nor do you have CROSS_COMPILE prefix when you do gcc -mno-cygwin (if available). But note that I'm *not* saying that this check is meaningless, formally one does need it. In sense that it's reasonable to assume that File::Spec->devnull() has no meaning to CROSS_COMPILE-ed binary. It's just that it doesn't cover all the corner cases. As already mentioned it's appropriate to `$^O eq 'MSWin32' or config($target) =~ m/^mingw/` in the beginning. Or maybe config($target) =~ m/^(VC|mingw)/ would be more explanatory...

Reply:
Once more, mingw perl has `$^O eq 'MSWin32'`, and this test is skipped on `MSWin32` (see higher up).

Reply:
> mingw perl has $^O eq 'MSWin32'...

ActiveState Perl has it, Straberry Perl has it. But when you build mingw on MSYS2, you're not using ActiveState or Straberry Perl, you're using MSYS2 perl, and it does *not* say MSWin32. When you build mingw on cygwin (yes, you can), you're using cygwin perl, and it does *not* say MSWin32. Sorry :-(

Reply:
i.e. `/mingw64/bin/perl` and `/mingw32/bin/perl`...

Reply:
Isn't cygwin kind of a small world of its own?

Reply:
> Isn't cygwin kind of a small world of its own?

And so is MSYS2. And one can argue that MSYS2 is smaller world of its own. But small or not you can use cygwin to build mingw. Well, now there are cross-compile packages just like for Linux, but originally you has *one* compiler and you could simply pass -mno-cygwin to generate plain Windows binaries with compilation facilitated by mingw [headers and libraries].

Reply:
> /mingw64/bin/perl and /mingw32/bin/perl...

Are there such things? Even if there are, does *everybody* install them? As one doesn't have to, and I for one don't... I just pacman -S perl mingw-w64-x86_64-gcc, and off it goes...

Reply:
Actually, mingw perl bad...  Because `$^O eq 'MSWIN32'`, Configure generates bad paths...

So back to msys perl.

On the other hand, cygwin passes these tests, including the `/dev/null` one, so I'm not sure why cygwin is a bother...

Reply:
Cygwin itself is not a problem, building [and testing] **mingw** on cygwin is.

Reply:
Do we support that?  I had no clue...

Reply:
BTW, I just tested this in a Mingw shell, mingw GNU toolchain and msys perl.  It passes...

Reply:
> Do we support that?

Well, it does work and it did work all alone. And that thing I said about -mno-cygwin, that's how I used to test mingw for many years (that's why `grep no-cygwin Configure` gives reading). I don't do that anymore, I simply do it on Linux nowadays. But the fact is that targeting mingw on cygwin is not different from targeting mingw on Linux, i.e. if it works on Linux, it will on Cygwin.

Reply:
> Cygwin itself is not a problem, building [and testing] **mingw** on cygwin is.

How does one even do that?  I just tried this in a 64-bit Cygwin shell:

```
./Configure mingw64; make -j8 test TESTS=test_out_option
```

It doesn't even build right.

Reply:
> mingw GNU toolchain and msys2 perl. It passes...

Grrrr! MSYS2 cheats! Remember that it tells apart binaries that are linked with MSYS2 run-time and binaries that are not? And magically translates /c/some/where command-line argument to c:\some\where when it executes the latter kind? Apparently it does same with /dev/null and translates it it nul...

Reply:
However! If the test in question passes on MSYS2, then why does one suppress it on '$^O eq MSWin32'? After all, mingw binary is effectively equivalent to one compiled with MSCV? After all both get linked with MSVCRT, so they share all those POSIX-y stat and company calls...

Reply:
> How does one even do that [build mingw on Cygwin]?

*Nowadays* you have to install cross-compile packages and process as on Linux, i.e. `./Configure mingw64 --cross-compile-prefix=x86_64-w64-mingw32-`. [An when -mno-cygwin was an option, it isn't anymore, ./Configure mingw just worked. But you still needed add-on packages tht comprised only headers and libraries.]

Reply:
> However! If the test in question passes on MSYS2, then why does one suppress it on '$^O eq MSWin32'?

One doesn't, unless one uses the mingw perl flavors (`/mingw64/bin/perl`), which are unusable.  So with msys perl (`/usr/bin/perl`), you get `$^O eq 'msys'`.

Reply:
> *Nowadays* you have to install cross-compile packages and process as on Linux, i.e. `./Configure mingw64 --cross-compile-prefix=x86_64-w64-mingw32-`. [~An~ And only when -mno-cygwin was an option, it isn't anymore, ./Configure mingw just worked. But you still needed add-on packages that comprised only headers and libraries.]

Could this please please please get documented in `NOTES.WIN`?  It's a *royal pain* having to remember all those special cases and details.

Reply:
> > If the test in question passes on MSYS2, then why does one suppress it on '$^O eq MSWin32'?

> One doesn't, unless one uses the mingw perl flavors

The question is about binary, not perl. My impression is that tests are suppressed because Windows code can't handle them. As already mentioned, mingw and MSVC effectively share run-time, so that if binary compiled by mingw compiler passes the test, then binary compiled by MSVC should do too.

Reply:
Mm, ok, I hear you...